### ISSUE TITLE
java.lang.IllegalStateException: Released

### ISSUE THREAD

**Issue by ollierob** [OpenHFT/Chronicle-Queue#479]
Recently upgraded from 4.6.77 -> 4.6.109 and started seeing this exception on occasion:
```
java.lang.IllegalStateException: Released: stack trace [java.lang.IllegalStateException: Released
	at net.openhft.chronicle.core.ReferenceCounter.reserve(ReferenceCounter.java:56)
	at net.openhft.chronicle.bytes.MappedFile.reserve(MappedFile.java:342)
	at net.openhft.chronicle.bytes.MappedBytes.reserve(MappedBytes.java:69)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:61)
	at net.openhft.chronicle.bytes.MappedBytes.<init>(MappedBytes.java:55)
	at net.openhft.chronicle.bytes.MappedBytes.mappedBytes(MappedBytes.java:112)
	at net.openhft.chronicle.queue.impl.single.ReferenceCountedCache.get(ReferenceCountedCache.java:44)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue$StoreSupplier.acquire(SingleChronicleQueue.java:787)
	at net.openhft.chronicle.queue.impl.WireStorePool.acquire(WireStorePool.java:97)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.storeForCycle(SingleChronicleQueue.java:404)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.cycle(SingleChronicleQueueExcerpts.java:1806)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.toStart(SingleChronicleQueueExcerpts.java:1541)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueue.createTailer(SingleChronicleQueue.java:398)
```
I am not caching/pooling tailers in any way, however there are multiple threads reading off the same queue instance.

Looking in the source code it seems possible and valid for `value=0` to occur for some short time in `release`, IMO you should only throw an exception if `value<0` - or better yet don't even check it inside `reserve`.

**Comment by RobAustin** [comment]
We have not seen this our selves,  could you create a JUnit test that reproduces this issue. It would help us fix the problem, - please submit it as a pull request.

**Comment by ollierob** [comment]
It seems to be a race condition between release and reserve from multiple threads, so perhaps easier to conclude on the cause and fix by looking at & reasoning about the code than attempting to test it - but I'll give it a stab.

**Comment by ollierob** [comment]
Wrote a unit test with a bunch of threads calling `reseve` and `release` but wasn't able to reproduce, I think it would need a very finely crafted test to have one thread `reverse` when the last calls `release`.

I would still maintain that a sensible fix here is to simply remove the `value.get()` call & assertion within the `reserve` method.

**Comment by JerryShea** [comment]
I can't reproduce this either. While I can see from the code **how** this could happen, can you explain what your code is doing and when & how often you see it please?

**Comment by ollierob** [comment]
I have one thread writing into a queue, multiple threads reading out of it. When the writer writes into the queue it subsequently submits a number of executor service jobs (typically `$numJobs > $numThreads`) which performs some work based on the recently written data (in our case, some "slow" DB writes). Each job starts by creating a tailer as per the original stack trace - and the rest is history.

I have only seen it a couple of times in the last few days - I am reading/writing millions of items/day - it is clearly quite rare, as you'd expect.

**Comment by RobAustin** [comment]
@ollierob We recently released 5.16.1 this changes that way we atomically write to the queue, although I appreciate you only have a single appender, could you retest and let me know if you still see this issue.

**Comment by ollierob** [comment]
Thanks Rob, but I still see this happening on occasion. Isn't this caused by a count of the readers hitting 0 when a new reader wants to read from the same file?

**Comment by ollierob** [comment]
This "released" issue seems to have been eclipsed by another read error in 5.16.6:

```
Caused by: java.nio.BufferOverflowException
	at net.openhft.chronicle.bytes.MappedBytes.acquireNextByteStore0(MappedBytes.java:363)
	at net.openhft.chronicle.bytes.MappedBytes.acquireNextByteStore(MappedBytes.java:352)
	at net.openhft.chronicle.bytes.MappedBytes.peekVolatileInt(MappedBytes.java:673)
	at net.openhft.chronicle.wire.AbstractWire.readDataHeader(AbstractWire.java:203)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.inACycle(SingleChronicleQueueExcerpts.java:1156)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:950)
	at net.openhft.chronicle.queue.impl.single.SingleChronicleQueueExcerpts$StoreTailer.readingDocument(SingleChronicleQueueExcerpts.java:892)
	at com.x.yRH$KjS.b(SourceFile:148)
	... 12 more
Caused by: java.io.IOException: Closed
	at net.openhft.chronicle.bytes.MappedFile.acquireByteStore(MappedFile.java:261)
	at net.openhft.chronicle.bytes.MappedFile.acquireByteStore(MappedFile.java:254)
	at net.openhft.chronicle.bytes.MappedBytes.acquireNextByteStore0(MappedBytes.java:358)
```

**Comment by hft-team-city** [comment]
Released in Chronicle-Queue-5.16.13 (https://github.com/OpenHFT/Chronicle-Queue/releases/chronicle-queue-5.16.13), BOM-1.16.170 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-1.16.170)

**Comment by ollierob** [comment]
Fantastic, will give it a spin today.
